### PR TITLE
Add Xds flow control service

### LINKED ISSUE #1515: Supports xDS protocol to communicate with the Control Plane of Service Mesh
# Value
Sermant is a Proxyless Service Mesh Solution with better performance.The support for the xDS protocol enables Sermant to better integrate into the Cloud Native Ecosystem. In Istio, Sermant can replace Envoy as the data plane to obtain directly the configuration of the control plane, and avoid performance loss and architecture complexity caused by Sidecar containers.
# Objective
Sermant can connect to the control plane of Istio through the xDS API interface to obtain the configurations of CDS, EDS, LDS, and RDS, and parse configurations to data model of the Sermant. Sermant will  implement gradually capabilities such as service discovery , route configuration and flow control, and the the host microservice can implements the Proxyless Mesh capability by mounting Sermant to  invoke directly the upstream service registered with the control plane, such as istiod.
# RoadMap
**v2.0.0 (2024/06/30)**
Supports xDS protocol for service discovery
**v2.1.0 (2024/09/30)**
Supports xDS protocol for routing and load balancing
**v2.2.0 (2024/12/30)**
Supports xDS protocol for flow control

### PR DESCRIPTION
**Pull request by hanbingleixue**
**What type of PR is this?**

Feature

**What this PR does / why we need it?**

Add Xds flow control service

**Which issue(s) this PR fixes？**

Fixes #1515

**Does this PR introduce a user-facing change?**

No

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/sermant-io/Sermant/issues) related with this PR before you start working on it.
- [x] Make sure you have squashed your change to one single commit.
- [ ] GitHub Actions works fine in this PR.

### CODE REVIEW

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/xds/XdsFlowControlService.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    Optional<XdsCircuitBreakers> getCircuitBreakers(String serviceName, String clusterName);
+
+    /**
+     * get Outlier Detection of cluster
+     *
+     * @param serviceName service name
+     * @param clusterName cluster name
+     * @return Outlier Detection rules
+     */
+    Optional<XdsOutlierDetection> getOutlierDetection(String serviceName, String clusterName);
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/xds/XdsFlowControlService.java]: CircuitBreaker should contain OutlierDetection

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/CdsProtocolTransformer.java [hunk 2]
```diff
@@ -113,4 +123,32 @@ private static String getServiceBaseClusterName(Set<XdsCluster> xdsClusters) {
...
+    private static XdsOutlierDetection parseClusterOutlierDetection(OutlierDetection outlierDetection) {
+        XdsOutlierDetection xdsOutlierDetection = new XdsOutlierDetection();
+        xdsOutlierDetection.setSplitExternalLocalOriginErrors(outlierDetection.getSplitExternalLocalOriginErrors());
+        UInt32Value localOriginFailure = outlierDetection.getConsecutiveLocalOriginFailure();
+        xdsOutlierDetection.setConsecutiveLocalOriginFailure(localOriginFailure.getValue());
+        UInt32Value gatewayFailure = outlierDetection.getConsecutiveGatewayFailure();
+        xdsOutlierDetection.setConsecutiveGatewayFailure(gatewayFailure.getValue());
+        UInt32Value consecutive5Xx = outlierDetection.getConsecutive5Xx();
+        xdsOutlierDetection.setConsecutiveGatewayFailure(consecutive5Xx.getValue());
+        Duration interval = outlierDetection.getInterval();
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/CdsProtocolTransformer.java]: no need temp variable

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/RdsProtocolTransformer.java [hunk 3]
```diff
@@ -108,6 +135,22 @@ private static XdsRoute parseRoute(Route route) {
         xdsRoute.setName(route.getName());
         xdsRoute.setRouteMatch(parseRouteMatch(route.getMatch()));
         xdsRoute.setRouteAction(parseRouteAction(route.getRoute()));
+        for (Map.Entry<String, Any> entry : route.getTypedPerFilterConfigMap().entrySet()) {
+            if (entry.getValue() == null) {
+                continue;
+            }
+            if (StringUtils.equals(entry.getKey(), "envoy.filters.http.fault")) {
```
Comment (zwmagic) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/RdsProtocolTransformer.java]: suggest setting it as a constant.

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/RdsProtocolTransformer.java [hunk 4]
```diff
@@ -215,4 +259,135 @@ private static XdsClusterWeight parseClusterWeight(ClusterWeight clusterWeight)
...
+            responseHeader.setHeader(xdsHeader);
+            responseHeaders.add(responseHeader);
+        }
+        return responseHeaders;
+    }
+
+    private static Optional<io.sermant.core.service.xds.entity.FractionalPercent> parseRuntimeFractionalPercent(
+            Struct filterEnabledStruct) {
+        Map<String, Value> valueMap = filterEnabledStruct.getFieldsMap();
+        if (valueMap.get("default_value") == null) {
```
Comment (zwmagic) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/RdsProtocolTransformer.java]: valueMap.get("default_value") can be assigned to a variable

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/xds/XdsFlowControlService.java (hunk 1 above)
Comment (hanbingleixue) [reply sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/xds/XdsFlowControlService.java]: done

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/CdsProtocolTransformer.java (hunk 2 above)
Comment (hanbingleixue) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/CdsProtocolTransformer.java]: done

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/RdsProtocolTransformer.java (hunk 3 above)
Comment (hanbingleixue) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/RdsProtocolTransformer.java]: done

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/RdsProtocolTransformer.java (hunk 4 above)
Comment (hanbingleixue) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/RdsProtocolTransformer.java]: done

**Review by zwmagic — APPROVED** [review]

**Review by zwmagic — DISMISSED** [review]

**Review by zwmagic — DISMISSED** [review]

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/xds/XdsFlowControlService.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+/**
+ * xDS FlowControl service
+ *
+ * @author zhp
+ * @since 2024-11-27
+ **/
+public interface XdsFlowControlService {
+    /**
+     * get circuit breaker information for downstream instance requests
```
Comment (lilai23) [inline sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/xds/XdsFlowControlService.java]: upstream, check others

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/RdsProtocolTransformer.java [hunk 6]
```diff
@@ -215,4 +260,138 @@ private static XdsClusterWeight parseClusterWeight(ClusterWeight clusterWeight)
         xdsClusterWeight.setClusterName(clusterWeight.getName());
         return xdsClusterWeight;
     }
+
+    private static XdsRetryPolicy parseRetryPolicy(RetryPolicy retryPolicy) {
+        XdsRetryPolicy xdsRetryPolicy = new XdsRetryPolicy();
+        xdsRetryPolicy.setRetryOn(retryPolicy.getRetryOn());
+        xdsRetryPolicy.setMaxAttempts(retryPolicy.getHostSelectionRetryMaxAttempts());
+        long perTryTimeout = java.time.Duration.ofSeconds(retryPolicy.getPerTryTimeout().getSeconds()).toMillis();
```
Comment (lilai23) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/RdsProtocolTransformer.java]: import java.time.Duration

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/RdsProtocolTransformer.java [hunk 7]
```diff
@@ -215,4 +260,138 @@ private static XdsClusterWeight parseClusterWeight(ClusterWeight clusterWeight)
...
+            return Optional.empty();
+        }
+        double tokensPerFill = fieldMap.get("tokens_per_fill").getNumberValue();
+        long fillInterval = Duration.parse("PT" + fieldMap.get("fill_interval").getStringValue()).toMillis();
+        XdsTokenBucket xdsTokenBucket = new XdsTokenBucket();
+        xdsTokenBucket.setFillInterval(fillInterval);
+        xdsTokenBucket.setMaxTokens((int) fieldMap.get("max_tokens").getNumberValue());
+        xdsTokenBucket.setTokensPerFill((int) tokensPerFill);
+        return Optional.of(xdsTokenBucket);
+    }
```
Comment (lilai23) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/RdsProtocolTransformer.java]: Too many strings needed to be managed. Besides I suggest you to give an example structure for these fileds

Comment (hanbingleixue) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/RdsProtocolTransformer.java]: done

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/RdsProtocolTransformer.java (hunk 6 above)
Comment (hanbingleixue) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/RdsProtocolTransformer.java]: done

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/xds/XdsFlowControlService.java (hunk 5 above)
Comment (hanbingleixue) [reply sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/xds/XdsFlowControlService.java]: done

**Review by lilai23 — APPROVED** [review]

**Review by daizhenyu — APPROVED** [review]

**Pull request merged**